core-bundle/tests/Twig/Interop/ContaoEscaperNodeVisitorTest.php

```diff
@@ -97,16 +99,47 @@ public function testUppercaseEntities(): void
         $this->assertSame('&quot;A&quot; &amp; &lt;B&gt;', $output);
     }
 
+    public function testHtmlAttrFilter(): void
+    {
+        $templateContent = '<span title={{ title|e(\'html_attr\') }}></span>';
+
+        $output = $this->getEnvironment($templateContent)->render(
+            'legacy.html.twig',
+            [
+                'title' => '{{flavor}} _is_ a flavor',
+            ]
+        );
+
+        $this->assertSame('<span title=vanilla&#x20;_is_&#x20;a&#x20;flavor></span>', $output);
+    }
+
     private function getEnvironment(string $templateContent): Environment
     {
         $loader = new ArrayLoader([
             'modern.html.twig' => $templateContent,
             'legacy.html.twig' => $templateContent,
         ]);
 
+        $controller = $this->getMockBuilder(Adapter::class)
+            ->disableOriginalConstructor()
+            ->addMethods(['replaceInsertTags'])
+            ->setMockClassName(Controller::class)
```

Comment:
I have no idea why this is necessary. A normal mock works (but has the wrong type in `getAdapter`) and setting the mock class name works. If I don't do either, the adapter is `null` when multiple tests are run.

IMHO looks a but like a PHPUnit caching issue.

Reply:
This triggers the following warnings on my system:

```
There were 5 warnings:

1) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testEscapesEntities
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

2) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testDoesNotDoubleEncode
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

3) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testHandlesFiltersAndFunctions
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

4) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testUppercaseEntities
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

5) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testHtmlAttrFilter
Class "Contao\Controller" already exists.
```

Reply:
Ah, then this might be an autoloading issue. :thinking:

Reply:
This was a tough one. Fixed in 4dd61be7b.

Generating and executing Twig sources did not play well with PHPUnits reference handling. I rewrote the code to use old school `Controller::replaceInsertTags()` instead and the tests to mock and set a container with `System::getContainer` and real insert tag listeners.